servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/OptionalSslTest.java

```diff
@@ -0,0 +1,115 @@
+import io.servicetalk.transport.netty.internal.ExecutionContextExtension;
+
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.net.InetSocketAddress;
+import java.nio.channels.ClosedChannelException;
+import java.util.stream.Stream;
+
+import static io.servicetalk.concurrent.api.Single.succeeded;
+import static io.servicetalk.http.api.HttpSerializers.textSerializerUtf8;
+import static io.servicetalk.http.netty.BuilderUtils.newClientBuilder;
+import static io.servicetalk.http.netty.BuilderUtils.newServerBuilder;
+import static io.servicetalk.test.resources.DefaultTestCerts.serverPemHostname;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+/**
+ * Verifies the behavior and functionality of {@link ServerSslConfig#acceptInsecureConnections()}.
+ */
+final class OptionalSslTest {
+
+    @RegisterExtension
+    static final ExecutionContextExtension SERVER_CTX =
+            ExecutionContextExtension.cached("server-io", "server-executor")
+                    .setClassLevel(true);
+    @RegisterExtension
+    static final ExecutionContextExtension CLIENT_CTX =
+            ExecutionContextExtension.cached("client-io", "client-executor")
+                    .setClassLevel(true);
+
+    @ParameterizedTest(name = "{displayName} [{index}] acceptEnabled={0}, protocol={1}")
+    @MethodSource("args")
+    void acceptsEncryptedAndNonEncryptedRequests(final boolean acceptEnabled, final HttpProtocol protocol)
+            throws Exception {
+        final HttpService service = (ctx, request, responseFactory) ->
+                succeeded(responseFactory.ok().payloadBody("Hello World!", textSerializerUtf8()));
+
+        try (ServerContext server = serverBuilder(acceptEnabled, protocol).listenAndAwait(service)) {
+            try (BlockingHttpClient client = clientBuilder(server, true, protocol).buildBlocking()) {
+                final HttpResponse response = client.request(client.get("/"));
+                assertEquals(HttpResponseStatus.OK, response.status());
+            }
+
+            try (BlockingHttpClient client = clientBuilder(server, false, protocol).buildBlocking()) {
+                final HttpRequest request = client.get("/");
+                if (acceptEnabled) {
+                    final HttpResponse response = client.request(request);
+                    assertEquals(HttpResponseStatus.OK, response.status());
+                } else {
+                    assertThrows(ClosedChannelException.class, () -> client.request(request));
+                }
+            }
+        }
+    }
+
+    private static Stream<Arguments> args() {
+        return Stream.of(
+                Arguments.of(true, HttpProtocol.HTTP_1),
+                Arguments.of(false, HttpProtocol.HTTP_1),
```

Comment:
We should also test both paths (plaintext and secure) for the following scenarios:
1. Different `SslProvier`(s)
2. TLSv1.3 / TLSv1.2
3. mTLS / no mTLS
4. ALPN / no ALPN
5. SecurityHandshakeObserver - we need to verify that observer is invoked when secure client makes a request and that it's not invoked when plaintext client makes a request.

You can either add more tests here for these scenarios or go to `SslProvidersTest`, `Tls13Test`, `MutualSslTest`, `AlpnClientAndServerTest`, `SecurityHandshakeObserverTest` to enhance those in a similar way you improved `SniTest`.